stake-pool/single-pool/tests/deposit.rs

```diff
@@ -0,0 +1,453 @@
+    let pool_stake_after = pool_stake_after.unwrap().delegation.stake;
+
+    let expected_deposit = if activate {
+        alice_stake_before_deposit
+    } else {
+        stake_lamports
+    };
+
+    // deposit stake account is closed
+    assert!(context
+        .banks_client
+        .get_account(alice_default_stake)
+        .await
+        .expect("get_account")
+        .is_none());
+
+    // stake moved to pool
+    assert_eq!(minimum_stake + expected_deposit, pool_stake_after);
+
+    // alice got her rent back if active, or only paid fees otherwise
+    assert_eq!(
+        wallet_lamports_after_deposit,
+        USER_STARTING_LAMPORTS - expected_deposit - fees
+    );
+
+    // alice got tokens. no rewards have been paid so tokens correspond to stake 1:1
+    assert_eq!(
+        get_token_balance(&mut context.banks_client, &accounts.alice_token).await,
+        expected_deposit,
+    );
+}
+
+#[test_case(true; "activated")]
+#[test_case(false; "activating")]
+#[tokio::test]
+async fn fail_uninitialized(activate: bool) {
+    let mut context = program_test().start_with_context().await;
+    let accounts = SinglePoolAccounts::default();
+    let stake_account = Keypair::new();
+
+    let first_normal_slot = context.genesis_config().epoch_schedule.first_normal_slot;
+    context.warp_to_slot(first_normal_slot).unwrap();
+
+    create_vote(
+        &mut context.banks_client,
+        &context.payer,
+        &context.last_blockhash,
+        &accounts.validator,
+        &accounts.vote_account,
+    )
+    .await;
+
+    let token_account =
+        atoken::get_associated_token_address(&context.payer.pubkey(), &accounts.mint);
+
+    create_independent_stake_account(
+        &mut context.banks_client,
+        &context.payer,
+        &context.last_blockhash,
+        &stake_account,
+        &Authorized {
+            staker: context.payer.pubkey(),
+            withdrawer: context.payer.pubkey(),
+        },
+        &Lockup::default(),
+        TEST_STAKE_AMOUNT,
+    )
+    .await;
+
+    delegate_stake_account(
+        &mut context.banks_client,
+        &context.payer,
+        &context.last_blockhash,
+        &stake_account.pubkey(),
+        &context.payer,
+        &accounts.vote_account.pubkey(),
+    )
+    .await;
+
+    if activate {
+        advance_epoch(&mut context).await;
+    }
+
+    let instructions = instruction::deposit(
+        &id(),
+        &accounts.vote_account.pubkey(),
+        &stake_account.pubkey(),
+        &token_account,
+        &context.payer.pubkey(),
+        &context.payer.pubkey(),
+    );
+    let message = Message::new(&instructions, Some(&context.payer.pubkey()));
+    let transaction = Transaction::new(&[&context.payer], message, context.last_blockhash);
+
+    // this gives a random borsh error upon attempting to deserialize the mint
+    // TODO perhaps in a separate pr, wrap this for all processors with a more helpful error
```

Comment:
That actually makes sense -- to save on compute, you're deserializing the stake account with Borsh in the program. Since the stake account is uninitialized, you'll get that Borsh error

Reply:
yea i just mean if im a user trying to deposit into a pool that doesnt exist i would see borsh garbage but would rather see "someone needs to initialize this pool"